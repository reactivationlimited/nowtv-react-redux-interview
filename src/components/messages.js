import React from 'react'

export const Messages = ({ status, messages}) => {
	switch(status) {
		case 'PENDING': return <div>Loading messages</div>
		case 'FETCHED': return <div>{
			messages.map(({id, userId, message, timestamp}) => <div key={id}>{message}</div>)
		}</div>
		default: return null;
	}
}