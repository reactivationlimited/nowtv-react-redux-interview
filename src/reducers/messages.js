import { MESSAGES_PENDING, MESSAGES_FETCHED } from '../actions'
import { getMessages, getMembers } from '../data/index.js'

export const messagesPending = () => ({
  type: MESSAGES_PENDING,
})

export const messagesFetched = (messages) => ({
  type: MESSAGES_FETCHED,
  messages,
})

export const fetchMessages = () => async (dispatch) => {
  dispatch(messagesPending())

  // In a production app, I'd create another reducer etc. for members
  const [ messages, members ] = await Promise.all([getMessages(), getMembers()]);

  // I wouldn't join the data here in a production app, I'd use a selector or join on the server side
  messages.forEach(message => message.member = members.find(({id}) => id === message.userId))
  dispatch(messagesFetched(messages))
}

export const reducer = (state = {
  messages: []
}, action = {}) => {
  switch (action.type) {
  	case MESSAGES_PENDING: return {
      status: 'PENDING',
      messages: []
  	}
  	case MESSAGES_FETCHED: return {
  		...state,
  		status: 'FETCHED',
  		messages: action.messages
  	}
    default:
      return state
  }
}
