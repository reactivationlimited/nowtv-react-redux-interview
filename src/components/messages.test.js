import React from 'react';
import { shallow } from 'enzyme';

import { Messages } from './messages'

describe('The Messages component', () => {
	describe('when the status is PENDING', () => {
		const wrapper = shallow(<Messages status="PENDING" messages="data"/>)

		it('should render a loading message', () => {
			expect(wrapper.contains('Loading messages')).toBe(true);
		})
		it('should not render messages', () => {
			expect(wrapper.contains('data')).toBe(false);
		})
		it('should match a snapshot', () => {
			expect(wrapper).toMatchSnapshot()
		})
	})
	describe('when the status is FETCHED', () => {
		const wrapper = shallow(<Messages status="FETCHED" messages="messages"/>)

		it('should render messages', () => {
			expect(wrapper.contains('messages')).toBe(true);
		})
		it('should not render a loading message', () => {
			expect(wrapper.contains('Loading messages')).toBe(false);
		})
		it('should match a snapshot', () => {
			expect(wrapper).toMatchSnapshot()
		})
	})
})