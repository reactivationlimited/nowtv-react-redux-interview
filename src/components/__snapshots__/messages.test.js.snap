// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Messages component when the status is FETCHED should match a snapshot 1`] = `
<div>
  <div
    key="cd445e6d-e514-424f-ba8f-16ec842002c6"
  >
    Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.
  </div>
  <div
    key="b03569ae-ccbf-4975-8040-4daba638b407"
  >
    Integer a nibh. In quis justo. Maecenas rhoncus aliquam lacus.
  </div>
</div>
`;

exports[`The Messages component when the status is PENDING should match a snapshot 1`] = `
<div>
  Loading messages
</div>
`;
